### PR TITLE
Add support for HIP-583

### PR DESCRIPTION
**Pull request by bamzedev**
Signed-off-by: Emanuel Pargov <bamzedev@gmail.com>

**Description**:

This PR adds support for [HIP-583](https://hips.hedera.com/hip/hip-583)

**Related issue(s)**:

Fixes https://github.com/hashgraph/hedera-sdk-go/issues/650, https://github.com/hashgraph/hedera-sdk-go/issues/637, https://github.com/hashgraph/hedera-sdk-go/issues/602, https://github.com/hashgraph/hedera-sdk-go/issues/653, https://github.com/hashgraph/hedera-sdk-go/issues/676

### CODE REVIEW

File: account_info.go [hunk 1]
```diff
@@ -64,12 +64,14 @@ func _AccountInfoFromProtobuf(pb *services.CryptoGetInfoResponse_AccountInfo) (A
 	if pb == nil {
 		return AccountInfo{}, errParameterNull
 	}
-
-	pubKey, err := _KeyFromProtobuf(pb.Key)
-	if err != nil {
-		return AccountInfo{}, err
+	var pubKey Key
+	if pb.Key.Key != nil {
```
[the remark this answers is not in the document]
Comment (bamzedev) [reply account_info.go:68]: When the account is hollow, yes.

File: account_id.go [hunk 2]
```diff
@@ -361,9 +366,9 @@ func (id AccountID) Compare(given AccountID) int {
 		}
 	}
 
-	if id.AliasEvmAddress != nil && given.AliasEvmAddress != nil {
-		originalEvmAddress := hex.EncodeToString(*id.AliasEvmAddress)
-		givenEvmAddress := hex.EncodeToString(*id.AliasEvmAddress)
+	if id.EvmAddress != nil && given.EvmAddress != nil {
+		originalEvmAddress := hex.EncodeToString(*id.EvmAddress)
+		givenEvmAddress := hex.EncodeToString(*id.EvmAddress)
```
[the remark this answers is not in the document]
Comment (bamzedev) [reply account_id.go:371]: Should be as is for now until https://github.com/hashgraph/hedera-services/issues/4606 is fixed.

File: account_id.go [hunk 3]
```diff
@@ -268,12 +273,12 @@ func _AccountIDFromProtobuf(accountID *services.AccountID) *AccountID {
 		initialKey, err := _KeyFromProtobuf(&pb)
```
[the remark this answers is not in the document]
Comment (bamzedev) [reply account_id.go:273]: Should be as is for now until https://github.com/hashgraph/hedera-services/issues/4606 is fixed.

File: account_id.go [hunk 4]
```diff
@@ -224,9 +229,9 @@ func (id AccountID) _ToProtobuf() *services.AccountID {
 		}
 
 		return resultID
-	} else if id.AliasEvmAddress != nil {
+	} else if id.EvmAddress != nil {
 		resultID.Account = &services.AccountID_Alias{
```
[the remark this answers is not in the document]
Comment (bamzedev) [reply account_id.go:233]: Should be as is for now until https://github.com/hashgraph/hedera-services/issues/4606 is fixed.

File: account_create_transaction.go [hunk 5]
```diff
@@ -367,8 +380,9 @@ func (transaction *AccountCreateTransaction) _ConstructScheduleProtobuf() (*serv
 	if transaction.aliasKey != nil {
 		key, _ := protobuf.Marshal(transaction.aliasKey._ToProtoKey())
 		body.Alias = key
-	} else if transaction.aliasEvmAddress != nil {
-		body.Alias = transaction.aliasEvmAddress
+	}
+	if transaction.evmAddress != nil {
+		body.Alias = transaction.evmAddress
```
[the remark this answers is not in the document]
Comment (bamzedev) [reply account_create_transaction.go:385]: Should be as is for now until https://github.com/hashgraph/hedera-services/issues/4606 is fixed.

**Review by MiroslavGatsanoga — APPROVED** [review]

File: account_create_transaction.go [hunk 6]
```diff
@@ -249,12 +249,14 @@ func (transaction *AccountCreateTransaction) GetAliasKey() PublicKey {
 	return PublicKey{}
 }
 
+// Deprecated - Use SetEvmAddress instead
```
Comment (MiroslavGatsanoga) [inline account_create_transaction.go:252]: The `Deprecated` comment is no longer relevant.

File: account_create_transaction.go [hunk 7]
```diff
@@ -249,12 +249,14 @@ func (transaction *AccountCreateTransaction) GetAliasKey() PublicKey {
...
 func (transaction *AccountCreateTransaction) SetAliasEvmAddress(evmAddress []byte) *AccountCreateTransaction {
 	transaction._RequireNotFrozen()
 	transaction.aliasEvmAddress = evmAddress
 	return transaction
 }
 
+// Deprecated - Use GetEvmAddress instead
```
Comment (MiroslavGatsanoga) [inline account_create_transaction.go:259]: The `Deprecated` comment is no longer relevant.

File: account_id.go [hunk 8]
```diff
@@ -48,57 +48,62 @@ type _AccountIDs struct { //nolint
 // AccountIDFromString constructs an AccountID from a string formatted as
 // `Shard.Realm.Account` (for example "0.0.3")
 func AccountIDFromString(data string) (AccountID, error) {
-	shard, realm, num, checksum, alias, aliasEvmAddress, err := _AccountIDFromString(data)
+	shard, realm, num, checksum, alias, AliasEvmAddress, err := _AccountIDFromString(data)
```
Comment (MiroslavGatsanoga) [inline account_id.go:51]: **nitpick:** ‏For consistency `AliasEvmAddress` should be in camel case as the rest of the variables.

File: account_id.go [hunk 9]
```diff
@@ -48,57 +48,62 @@ type _AccountIDs struct { //nolint
...
 		AliasEvmAddress: nil,
-		checksum:        checksum,
+		checksum:   checksum,
 	}, nil
 }
 
-func AccountIDFromEvmAddress(shard uint64, realm uint64, evmAddress string) (AccountID, error) {
-	temp, err := hex.DecodeString(evmAddress)
+func AccountIDFromEvmAddress(shard uint64, realm uint64, aliasEvmAddress string) (AccountID, error) {
+	temp, err := hex.DecodeString(aliasEvmAddress)
```
Comment (MiroslavGatsanoga) [inline account_id.go:89]: **suggestion:** ‏An additional check that the decoded value is 20 bytes can be added as well.

File: status.go [hunk 10]
```diff
@@ -313,6 +313,8 @@ const (
 	StatusInsufficientBalancesForStorageRent                       Status = 327
 	StatusMaxChildRecordsExceeded                                  Status = 328
 	StatusInsufficientBalancesForRenewalFees                       Status = 329
+	StatusTransactionHasUnknownFields                              Status = 330
+	StatusAccountIsImmutable                                       Status = 331
 )
```
Comment (MiroslavGatsanoga) [inline status.go:318]: **thought:** ‏Just to mention that there will be also [ALIAS_ALREADY_ASSIGNED](https://github.com/hashgraph/hedera-protobufs/blob/717b96c1ecf3fe5f9f30fe2bcbc731cf02934aae/services/response_code.proto#L1375) status as part of HIP-583, not sure when we will need to add it to the SDK as well.

**Pull request merged**